src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -295,6 +325,113 @@ void ScheduleTask(DateTimeOffset utcNow)
             }
         }
 
+        private readonly List<WeakReference<Stats>?>? _allStats;
+        private readonly Stats? _accumulatedStats;
+        private ThreadLocal<Stats>? _stats;
+
+        private (long, long) Sum()
+        {
+            long hits = 0, misses = 0;
+            lock (_allStats!)
+            {
+                if (IntPtr.Size == 4)
+                {
+                    Interlocked.Add(ref hits, Interlocked.Read(ref _accumulatedStats!.Hits));
+                    Interlocked.Add(ref misses, Interlocked.Read(ref _accumulatedStats!.Misses));
+                }
+                else
+                {
+                    hits += _accumulatedStats!.Hits;
+                    misses += _accumulatedStats!.Misses;
+                }
+
+                foreach (WeakReference<Stats>? wr in _allStats!)
+                {
+                    if (wr is not null && wr.TryGetTarget(out Stats? stats))
+                    {
+                        hits += Interlocked.Read(ref stats.Hits);
+                        misses += Interlocked.Read(ref stats.Misses);
+                    }
+                }
+            }
+            return (hits, misses);
+        }
+
+        private void Hit()
+        {
+            if (_options.TrackStatistics)
+            {
+                if (IntPtr.Size == 4)
+                    Interlocked.Increment(ref GetStats().Hits);
+                else
+                    GetStats().Hits++;
+            }
+        }
+
+        private void Miss()
+        {
+            if (_options.TrackStatistics)
+            {
+                if (IntPtr.Size == 4)
+                    Interlocked.Increment(ref GetStats().Misses);
+                else
+                    GetStats().Misses++;
+            }
+        }
+
+        private Stats GetStats() => _stats!.Value!;
+
+        internal sealed class Stats
+        {
+            public long Hits;
+            public long Misses;
+            private readonly MemoryCache _memoryCache;
+            public Stats(MemoryCache memoryCache, bool skipAdd = false)
+            {
+                _memoryCache = memoryCache;
+                if (!skipAdd)
+                    _memoryCache.AddToStats(this);
+            }
+            ~Stats()
+            {
+                _memoryCache.RemoveFromStats(this);
+            }
+        }
+
+        private void RemoveFromStats(Stats current)
+        {
+            lock (_allStats!)
+            {
+                for (int i = 0; i < _allStats!.Count; i++)
+                {
+                    if (_allStats![i] is WeakReference<Stats> wr && wr.TryGetTarget(out Stats? stats) && stats == current)
+                    {
+                        _allStats!.RemoveAt(i);
+                        break;
+                    }
+                }
+
+                if (IntPtr.Size == 4)
+                {
+                    Interlocked.Add(ref _accumulatedStats!.Hits, Interlocked.Read(ref current!.Hits));
```

Comment:
```suggestion
                    _accumulatedStats!.Hits += Interlocked.Read(ref current.Hits);
                    _accumulatedStats!.Misses += Interlocked.Read(ref current.Misses);
```
accumulatedStats are never modified outside the lock so no interlocked is necessary there.

Reply:
would I still not need interlocked because of preventing torn state on 32bit?

Reply:
Do we even need the `Interlocked.Read` in this case where it's only called from `current`'s finalizer? That means there can't be any thread concurrently modifying this, right?

I don't think we need to worry about any 32 bit tearing if there's no concurrent access and all writes have been committed which I would expect to be the case in a finalizer.

Reply:
The only other place the Misses and Hits are changed for Stats is in `Hit()` and `Miss()`

Reply:
OTOH, if it's a finalizer of a thread-local object, then I expect it will be called so rarely that hardly any CPU time will be spent there.

Reply:
Why is Interlocked.Read necessary on a long when in a 64 bit process? Can't we make it conditional on IntPtr.Size == 4?

Reply:
Although it isn't necessary it also shouldn't be problematic because this code isn't on the hot path. I doubt we'd see any perf benefit in exchange for adding the conditional.